src/ips.rs

```diff
@@ -0,0 +1,525 @@
+    pub async fn generate(&mut self, state: &CrunchyState, agraph: &AGraph) -> Vec<Peer> {
+        let mut peer_list = Vec::new();
+
+        // Reconstruct graph from the agraph - we need to do this because we need all the
+        // measures provided by spectre's graph implementation.
+        // Using agraph gives us certainity that we are using the same graph as the crawler and
+        // there are only good nodes there (this is critical assumption!). Second assumption is
+        // that agraph node indexes are the same as in the state.nodes vector.
+        let mut graph = self.construct_graph(&state.nodes, agraph);
+
+        // 0 - Detect islands
+        // To reconsider if islands should be merged prior to any other computations or not.
+        // IMHO, if there are islands they can influence on the results of the computations.
+        // TODO(asmie): Merging islands is not implemented yet.
+        let _islands = self.detect_islands(agraph);
+
+        // Now take the current params
+        let degrees = graph.degree_centrality();
+        let degree_avg = self.degree_centrality_avg(&degrees);
+        let eigenvalues = graph.eigenvalue_centrality();
+
+        // Determine factors used for normalization.
+        // Normalization step is needed to make sure that all the factors are in the same range and
+        // weights can be applied to them.
+        self.degree_factors =
+            NormalizationFactors::determine(&degrees.values().cloned().collect::<Vec<u32>>());
+
+        self.eigenvector_factors =
+            NormalizationFactors::determine(&eigenvalues.values().cloned().collect::<Vec<f64>>());
+
+        let betweenness = &state
+            .nodes
+            .iter()
+            .map(|n| n.betweenness)
+            .collect::<Vec<f64>>();
+        self.betweenness_factors = NormalizationFactors::determine(betweenness);
+
+        let closeness = &state
+            .nodes
+            .iter()
+            .map(|n| n.closeness)
+            .collect::<Vec<f64>>();
+        self.closeness_factors = NormalizationFactors::determine(closeness);
+
+        // Node rating can be split into two parts: constant and variable depending on the node's
+        // location. Now we can compute each node's constant rating based on some graph params.
+        // Vector contains IpAddr, node index (from the state.nodes) and rating. We need index just
+        // to be able to easily get the node from nodes vector after sorting.
+        let mut const_factors = Vec::with_capacity(state.nodes.len());
+        for node_idx in 0..state.nodes.len() {
+            let ip = IpAddr::from_str(state.nodes[node_idx].ip.as_str()).unwrap();
+            const_factors.push((
+                ip,
+                node_idx,
+                self.rate_node(
+                    &state.nodes[node_idx],
+                    *degrees.get(&ip).unwrap(), // should be safe to unwrap here
```

Comment:
> should be safe to unwrap here

Nit: I'd add why and make it obvious with e.g. `Safety: ...`

Reply:
Perhaps using `expect` with nice error messages is a good way to go, since there as a lot of `unwraps` based on similar assumptions in this code.

Reply:
Added expect error messages and refactored degree usage a bit later to avoid many unwraps.